src/poetry/puzzle/provider.py

```diff
@@ -469,7 +469,14 @@ def complete_package(
                 if extra not in package.extras:
                     continue
 
-                optional_dependencies += [d.name for d in package.extras[extra]]
+                for extra_dependency in package.extras[extra]:
+                    if extra_dependency.name == dependency.name:
+                        # If an extra/feature refers to another in the base package,
+                        # resolve the dependencies of that extra/feature.
+                        for feature in extra_dependency.features:
+                            for pd in package.extras.get(feature, []):
```

Comment:
What if the self-reference goes a layer deeper still? I think this wants to be a depth- first search (or some other graph traversal)